File: BinarySearch/BinarySearch.py
# https://leetcode.com/problems/binary-search/description/
def search(nums: list, target: int) -> int:
    if nums[0] == target:
        return 0
    left = 0
    right = len(nums) - 1
    while left <= right:
        mid = int((left+right) / 2)
        if target < nums[mid]:
            right = mid - 1
        elif target > nums[mid]:
            left = mid + 1
        else:
            return mid
    return -1

print(search([1,2,3,4,5,6,7,8,9,10], 5)) # 4
print(search([-1,0,3,5,9,12], 9)) # 4
print(search([-1,0,3,5,9,12], 2)) # -1
print(search([5], 5)) # 0
print(search([2,5], 5)) # 1
print(search([2,5], 2)) # 0

File: BinarySearch/BinarySearch.java
package BinarySearch;

public class BinarySearch {
    public int search(int[] nums, int target) {
        int begin = 0;
        int end = nums.length-1;
        while (end > begin) {
            int mid = (begin + end) / 2;
           if (nums[mid] == target) {
               return mid;
           } else if (nums[mid] < target) {
               begin = mid + 1;
           } else if (nums[mid] > target) {
               end = mid - 1;
           }
        }
        return -1;
    }
}